### PR TITLE
win midi: bring back unprepare header

### PR DESCRIPTION
**Pull request by rfomin**

### CODE REVIEW

File: src/i_winmusic.c [hunk 1]
```diff
@@ -199,49 +200,43 @@ static void CALLBACK MidiStreamProc(HMIDIOUT hMidi, UINT uMsg,
...
+
+#ifndef __SANITIZE_ADDRESS__
+    mmr = midiOutUnprepareHeader((HMIDIOUT)hMidiStream, hdr, sizeof(MIDIHDR));
     if (mmr != MMSYSERR_NOERROR)
     {
-        MidiError("midiOutPrepareHeader", mmr);
+        MidiError("midiOutUnprepareHeader", mmr);
     }
+#endif
+
```
Comment (ceski-1) [inline src/i_winmusic.c:231]: To be safe, add something like `memset(hdr, 0, sizeof(*hdr));` here or, at the very least, in `I_WIN_ShutdownMusic()`. This ensures the midi header is completely reset.

File: src/i_winmusic.c [hunk 2]
```diff
@@ -199,49 +200,43 @@ static void CALLBACK MidiStreamProc(HMIDIOUT hMidi, UINT uMsg,
...
+        return;
+    }
+
+    // Avoid ASan detection. Commentary by Microsoft: "It looks like
+    // midiOutPrepareHeader() allocates with HeapAlloc(), and then
+    // midiOutUnprepareHeader() deallocates with GlobalFree(GlobalHandle
+    // (...)). By design, this kind of allocator mismatch is an issue that ASan
+    // is designed to catch. It is theoretically possible for us to support
+    // this kind of code, but it’s not very high priority since it is undefined
+    // behavior, though it happens to work right now outside of ASan."
```
[the remark this answers is not in the document]
Comment (rfomin) [reply src/i_winmusic.c:222]: Fixed in https://github.com/fabiangreffrath/woof/pull/1268/commits/e7ac4c69220462a3e6b456e2f8f52d6b48632038

File: src/i_winmusic.c [hunk 3]
```diff
@@ -1877,6 +1873,9 @@ static void I_WIN_ShutdownMusic(void)
     }
     hMidiStream = NULL;
 
+    free(buffer.data);
```
[the remark this answers is not in the document]
Comment (rfomin) [reply src/i_winmusic.c:1878]: Fixed in https://github.com/fabiangreffrath/woof/pull/1268/commits/e7ac4c69220462a3e6b456e2f8f52d6b48632038

File: src/i_winmusic.c [hunk 4]
```diff
@@ -279,6 +276,15 @@ static void StreamOut(void)
 
     hdr->lpData = (LPSTR)buffer.data;
     hdr->dwBytesRecorded = buffer.position;
+    hdr->dwBufferLength = buffer.size;
```
[the remark this answers is not in the document]
Comment (rfomin) [reply src/i_winmusic.c:281]: I nullified the header like ZMusic does.

File: src/i_winmusic.c (hunk 1 above)
Comment (rfomin) [reply src/i_winmusic.c:232]: I decided to do the `memset` in `StreamOut()` instead, that's what ZMusic does. After looking at the `MIDIHDR` structure, I'm not sure if we should nullify it:
```c
/* MIDI data block header */
typedef struct midihdr_tag {
    LPSTR       lpData;               /* pointer to locked data block */
    DWORD       dwBufferLength;       /* length of data in data block */
    DWORD       dwBytesRecorded;      /* used for input only */
    DWORD_PTR   dwUser;               /* for client's use */
    DWORD       dwFlags;              /* assorted flags (see defines) */
    struct midihdr_tag far *lpNext;   /* reserved for driver */
    DWORD_PTR   reserved;             /* reserved for driver */
#if (WINVER >= 0x0400)
    DWORD       dwOffset;             /* Callback offset into buffer */
    DWORD_PTR   dwReserved[8];        /* Reserved for MMSYSTEM */
#endif
} MIDIHDR, *PMIDIHDR, NEAR *NPMIDIHDR, FAR *LPMIDIHDR;
```
It's from `mmeapi.h` in Windows SDK.

Comment (ceski-1) [reply src/i_winmusic.c:232]: Alternatively, could `MidiStreamHdr` be changed to a pointer? Seems weird to carry over a potentially dirty `MIDIHDR` from one device to the next.

Rough example:
```c
// Replaces MidiStreamHdr:
static MIDIHDR *hdr;

// I_WIN_InitMusic():
hdr = calloc(1, sizeof(*hdr));

// I_WIN_ShutdownMusic():
free(hdr);
hdr = NULL;
```

File: src/i_winmusic.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (ceski-1) [reply src/i_winmusic.c:281]: Good idea, that's even better.

File: src/i_winmusic.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (ceski-1) [reply src/i_winmusic.c:1878]: A few lines up, `ResetBuffer()` is called. If `buffer.prepared` is false, `buffer.position` isn't reset back to zero. I'm not sure if that situation is possible though. Add `buffer.position = 0;` to be safe?

[the remark this answers is not in the document]
Comment (rfomin) [reply src/i_winmusic.c:1879]: Fixed in https://github.com/fabiangreffrath/woof/pull/1268/commits/3dede7933840d21f94be5da19257baf8e6ec73a8

File: src/i_winmusic.c (hunk 1 above)
Comment (rfomin) [reply src/i_winmusic.c]: We don't carry it, we nullify it before using it in the `StreamOut()` function.
ZMusic has 100 times more users, so I think it's reasonable to copy their approach. However, GZDoom has long used Fluidsynth by default, so it's hard to tell how many users use midiStream() in ZMusic.

Comment (ceski-1) [reply src/i_winmusic.c]: You're right, that approach is fine.

**Review by ceski-1 — APPROVED** [review]
Would you like to make a PR for Chocolate before merging this one? Just in case there are additional changes requested.

**Comment by rfomin** [comment]
> Would you like to make a PR for Chocolate before merging this one? Just in case there are additional changes requested.

Ok, will do in a couple of hours.

**Pull request merged**